### PR TITLE
Refactoring. Moved common code from harvest and registry manager projects to commons.

### PR DESCRIPTION
**Pull request by tdddblog**
## 🗒️ Summary
Moved common code from harvest (standalone and web services projects) and registry manager project to commons.

## ⚙️ Test Data and/or Report
Test Harvest  (standalone and web services projects)

## ♻️ Related Issues
https://github.com/NASA-PDS/harvest/issues/78
https://github.com/NASA-PDS/harvest/issues/54

### CODE REVIEW

File: src/main/java/gov/nasa/pds/registry/common/dd/Pds2EsDataTypeMap.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        if(file == null) return;
+        
+        log.info("Loading PDS to ES data type mapping from " + file.getAbsolutePath());
+        
+        BufferedReader rd = null;
+        
+        try
+        {
+            rd = new BufferedReader(new FileReader(file));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/Pds2EsDataTypeMap.java:99]: ignore

File: src/main/java/gov/nasa/pds/registry/common/dd/parser/BaseLddParser.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    
+    /**
+     * Constructor
+     * @param file PDS LDD JSON file to parse
+     * @throws Exception an exception
+     */
+    public BaseLddParser(File file) throws Exception
+    {
+        this.ddFile = file;
+        jsonReader = new JsonReader(new FileReader(file));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/parser/BaseLddParser.java:37]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+     * @param is input stream
+     * @return Last line
+     */
+    private static String getLastLine(InputStream is)
+    {
+        String lastLine = null;
+
+        try
+        {
+            BufferedReader rd = new BufferedReader(new InputStreamReader(is));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:442]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+        
+        try
+        {
+            ZipEntry ze = zip.getEntry(fileName);
+            if(ze == null) 
+            {
+                throw new Exception("Could not find " + fileName +  " in " + zipFile.getAbsolutePath());
+            }
+            
+            BufferedReader rd = new BufferedReader(new InputStreamReader(zip.getInputStream(ze)));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:106]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+    /**
+     * Load data from an NJSON (new-line-delimited JSON) file into Elasticsearch.
+     * @param file NJSON (new-line-delimited JSON) file to load
+     * @throws Exception an exception
+     */
+    public void loadFile(File file) throws Exception
+    {
+        log.info("Loading ES data file: " + file.getAbsolutePath());
+        
+        BufferedReader rd = new BufferedReader(new FileReader(file));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:81]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/dd/GetDataTypesResponseParser.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+     * @param entity HTTP response
+     * @return a list of records
+     * @throws IOException an exception
+     */
+    public List<Record> parse(HttpEntity entity) throws IOException
+    {
+        List<Record> records = new ArrayList<>();
+        
+        InputStream is = entity.getContent();
+        rd = new JsonReader(new InputStreamReader(is));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/dd/GetDataTypesResponseParser.java:68]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/schema/MappingsParser.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Parse Elasticsearch response from "/indexName/_mappings" API.
+     * @param entity HTTP response body
+     * @return a collection of field names from a given Elasticsearch index.
+     * @throws IOException an exception
+     */
+    public Set<String> parse(HttpEntity entity) throws IOException
+    {
+        InputStream is = entity.getContent();
+        rd = new JsonReader(new InputStreamReader(is));
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/schema/MappingsParser.java:47]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/service/CollectionInventoryWriter.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    {
+        batch.batchNum = 0;
+        writer.clearData();
+        
+        InventoryBatchReader rd = null;
+        int writeCount = 0;
+        
+        try
+        {
+            rd = new InventoryBatchReader(new FileReader(inventoryFile), refType);
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/CollectionInventoryWriter.java:92]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/json/BaseNJsonWriter.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    
+    
+    /**
+     * Constructor
+     * @param file output file
+     * @throws Exception an exception
+     */
+    public BaseNJsonWriter(File file, String action) throws Exception
+    {
+        writer = new FileWriter(file);
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/json/BaseNJsonWriter.java:34]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    private DataDictionaryDao ddDao;
+    private SchemaDao schemaDao;
+    
+    // Use string data type for undefined fields
+    private boolean stringForMissing = true;
+    
+    /**
+     * Constructor
+     * @param cfg Registry (Elasticsearch) configuration
+     * @throws Exception
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java:43]: ignore

File: src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    private int numRetries = 3;
+    
+    private CloseableHttpClient httpClient;
+    
+
+    /**
+     * Constructor
+     * @param sslTrustAll Enable or disable SSL certificate and host validation 
+     * to support self-signed certificates.
+     * @throws Exception
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java:50]: ignore

File: src/main/java/gov/nasa/pds/registry/common/dd/Pds2EsDataTypeMap.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        
+        try
+        {
+            String line;
+            while((line = rd.readLine()) != null)
+            {
+                line = line.trim();
+                if(line.startsWith("#") || line.isEmpty()) continue;
+                String[] tokens = line.split("=");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/Pds2EsDataTypeMap.java:113]: ignore

File: src/main/java/gov/nasa/pds/registry/common/dd/parser/AttributeDictionaryParser.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+        
+        jsonReader.beginObject();
+        
+        while(jsonReader.hasNext() && jsonReader.peek() != JsonToken.END_OBJECT)
+        {
+            String name = jsonReader.nextName();
+            if("identifier".equals(name))
+            {
+                attr.id = jsonReader.nextString();
+                String tokens[] = attr.id.split("\\.");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/parser/AttributeDictionaryParser.java:91]: ignore

File: src/main/java/gov/nasa/pds/registry/common/dd/parser/ClassAttrAssociationParser.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        jsonReader.beginObject();
+        
+        while(jsonReader.hasNext() && jsonReader.peek() != JsonToken.END_OBJECT)
+        {
+            String name = jsonReader.nextName();
+            if("identifier".equals(name))
+            {
+                String id = jsonReader.nextString();
+                
+                String tokens[] = id.split("\\.");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/parser/ClassAttrAssociationParser.java:101]: ignore

File: src/main/java/gov/nasa/pds/registry/common/meta/InventoryBatchReader.java [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        batch.clear();
+        batch.batchNum++;
+        
+        String line;
+        int count = 0;
+        
+        while((line = rd.readLine()) != null)
+        {
+            if(line.isBlank()) continue;
+            String[] tokens = line.split(",");
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/meta/InventoryBatchReader.java:67]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/schema/SchemaDao.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Elasticsearch schema DAO (Data Access Object).
+ * This class provides methods to read and update Elasticsearch schema.
+ * 
+ * @author karpenko
+ */
+public class SchemaDao
+{
+    private Logger log;
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/schema/SchemaDao.java:23]: ignore

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 17]
```diff
@@ -0,0 +1,461 @@
...
+            if(responseHasErrors(respJson))
+            {
+                throw new Exception("Could not load data.");
+            }
+            
+            totalRecords += numRecords;
+
+            return line1;
+        }
+        catch(UnknownHostException ex)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:220]: fixed

File: src/main/java/gov/nasa/pds/registry/common/dd/LddEsJsonWriter.java [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            writeRecord(classNs, className, attr);
+        }
+    }
+
+    
+    /**
+     * Write PDS LDD version and date
+     * @param namespace LDD namespace
+     * @param version LDD version
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/dd/LddEsJsonWriter.java:92]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+                String file = (String)map.get("attr_name");
+                info.addSchemaFile(file);
+            }
+        }
+    }
+    
+
+    /**
+     * Get LDD date from data dictionary index in Elasticsearch.
+     * @param indexName Elasticsearch base index name, e.g., "registry".
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java:82]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+    private Pds2EsDataTypeMap dtMap;
+    private DataLoader loader;
+
+    private DataDictionaryDao dao;
+    
+    /**
+     * Constructor
+     * @param esUrl Elasticsearch URL
+     * @param esIndex Elasticsearch index name
+     * @param esAuthFile Elasticsearch authentication configuration file
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java:40]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+    
+    private Pds2EsDataTypeMap dtMap;
+    private DataLoader loader;
+
+    private DataDictionaryDao dao;
+    
+    /**
+     * Constructor
+     * @param esUrl Elasticsearch URL
+     * @param esIndex Elasticsearch index name
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java:39]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+
+    private SchemaUpdater schemaUpdater;
+    private FieldNameCache fieldNameCache;
+    
+    
+    /**
+     * Constructor.
+     * NOTE: Init registry manager before calling this constructor
+     * 
+     * @param cfg registry configuration
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java:31]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+     * @param ddDao data dictionary DAO
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java:32]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+     * @param sDao schema DAO
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/service/MissingFieldsProcessor.java:33]: fixed

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 25]
```diff
@@ -0,0 +1,461 @@
...
+            throw new Exception("Unknown host " + conFactory.getHostName());
+        }
+        catch(IOException ex)
+        {
+            // Get HTTP response code
+            int respCode = getResponseCode(con);
+            if(respCode <= 0) throw ex;
+            
+            // Try extracting JSON from multi-line error response (last line) 
+            String json = getLastLine(con.getErrorStream());
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:231]: ignore

That connection is cached by Java network library. Open HTTP connection is required for "keep-alive" feature to work.
`createConnection()` returns existing connection if it was created before.

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 26]
```diff
@@ -0,0 +1,461 @@
...
+            // Check for Elasticsearch errors.
+            String respJson = DaoUtils.getLastLine(con.getInputStream());
+            log.debug(respJson);
+            
+            int numErrors = processErrors(respJson);
+            // Return number of successfully saved records
+            // NOTE: data list has two lines per record (primary key + data)
+            return data.size() / 2 - numErrors;
+        }
+        catch(UnknownHostException ex)
```
[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:285]: fixed

[the remark this answers is not in the document]
Comment (tdddblog) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java:285]: ignore

**Review by tloubrieu-jpl — APPROVED** [review]
The tests I made worked, thanks @tdddblog

**Pull request merged**